### ISSUE TITLE
Bug: [unbound-method] Should error on call site instead of callee site

### ISSUE THREAD

**Issue by valler** [typescript-eslint/typescript-eslint#10318]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.tsx&code=PTAEGMEMBtoUzqAzgSwC5wDSgLaQJ4BGikAdvgO4AWcATiaQCagD2Armqx6wGYQuk0tFtABQ4AUk48WLUAF5QAb1GhQhSLQAUASmWq1oEKFJyABmiookZ7HAAeABzjhOp0HWG0DAX0wGNAC9dfUNQejQ2WlJQS2sAOgA3GABuIzBNAHM2SEJofA8nF05PFlpQGnpsAUQoGJQ8TJRSRCRIHjpQBsd4HDhBSDQUSV9-NWI4AC5QEPkAPlDDY1o2eFBGFjgkExZOSEcegrQ5QnYmXDhLFkYkeK6cHrg%2BwThmY67YNilaQdqYaGQ6DgowMyWg0wADP4fClRKJjChGMM8ENwKAvpBMogHJAHvAkHDjJA9FBYICMNhGFshGxXFFmpkIP8NOAANZISbiSScFTjTTYYylcpaNgtIquV46AKQQICsBCmaihzOCWMKXjBBynYeWheGbK4qS0Q%2BBSgGQsWGiDTaHSwoK6O0IB2EsCEPQAQQACgBJdFITqoCnIXEkbakgGBqbwsDh0AARlE%2BxQWnN2GtaZlaYQtpdTLJACZE45kwZU6XZPF0%2BkdXqRWKVRg1eWWJXM9WFXWDar1WaKxNq%2B4O13G1Kc9G8wD89NB7qykhKU8WNtqINQJALkhtpZatdEMQJH0CUmtM2xtWJKKMLRmpF0ChEogKD8DgzgIRmkjSJks1AvogWHwAGgBYVg2OoH4Mmm3AwEgLAGMY%2B6QH%2BsSgV0MSRNE4RwAAjmwKD0NseQsL8zBIPgAz2GhTL%2BrwoCJCwiITiy7KcmoswLLyhjmq2NqwmofjwTGZxXjeeFDA%2BoBPvsxZfm%2BEFfqxMx6PMvYtvaOhngigEtJJlSIO8MiwCwFAoYgjiDFeMTsJwQFULk6DYDgZQ7lSyB9GSUiaGgBJqMYrIoGyaF3O4LSvLEchSY46gIIp7GLGo3ETM6-GiGOxjgB6Pr3I8zzHOUkaOQQxBruQ1B0AwzDWbREiCMIYiCuKQxfiVrCEAAVsU2W9P0TWMmQaFXjwkDgBVhQNgyTHDeyog8KKrjDDEx4GCw7XTEo6iaLo0z0YxAlqOAhAQtMcU7YwZ4HXGx3KQspBsDgxC0OdhBTkpCgLKdqWLCtbU8clECHX9F2A89zo%2BEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEA9LWOkngJ7hgC%2BIHQA&tokens=false

### Repro Code

```TypeScript
// callee site, maybe anywhere and out out of control
const foo = {
  bar() {
    // no `this`, expect no error
  },
  baz() {
    return this.val; // arguably expect error here, one can imagine safer implementations
  },
  bee: () => {
    // rule does not apply to bound methods. implemented to illustrate call site
  },
  val: 0,
};

// idiomatic usage examples

// a) call site, destructuring callbacks:
const {
  bar, // error (unexpected)
  baz, // error (unexpected)
  bee, // no error (expected)
} = foo;

bar();
baz();
bee();

// b) API user site, same as call site:
// call 1
api(foo, bar, baz, bee);

// call 2
api(
  foo,
  foo.bar, // error (unexpected)
  foo.baz, // error (unexpected)
  foo.bee // no error (expected)
);

// call 2: no errors, demos what a mess the code becomes
api(
  foo,
  // counterintuitive wrapping/binding, because of of `this` binding, but also
  // because this in turn requires bloated syntax in case of void callbacks:
  () => {
    foo.bar();
  },
  // counterintuitive wrapping/binding:
  () => foo.baz(),
  // if one where to follow the pattern out of habit, more code smell starts
  // kick in. no need to wrap bee:
  () => {
    foo.bee();
  }
);

// c) API implementor site, maybe anywhere and out of control
// expecting an object implementing an interface and expecting callbacks
function api(
  obj: { bar(): void },
  cb0: () => void,
  cb1: () => number,
  cb2: () => void
) {
  obj.bar();
  cb0();
  cb1();
  cb2();
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // empty
  }
}
```

### Expected Result

I would expect what the rule suggests "Enforce unbound methods are called with their expected scope."

The enforcing and error reporting should happen on the callee site, where the actual error occurs, because of  (supposedly) unsafe usage of `this`.

### Actual Result

The enforcing and error happen on the call/user site. The problems of which are described in my original comment given in the additional info and illustrated by example.

### Additional Info

Can someone explain why this rule errors on the call site?

This already bit me more than once with false positives when working with external packages. My thoughts on this are as follows.

1) The suggestion to let the caller bind or wrap any unbound function or method for usage as a callback kinda advocates code smell (one implementation, up to infinity call sites), while passing around functions is perfectly idiomatic JS/TS (debatable).
2) Giving the call site tipps about improvements for the callee doesn't help at all, because as already mentioned, one might not be in control.
3) The caller should only be concerned about matching signatures and passing correct arguments. Most certainly the caller should not care about where and how a callable was defined, other than constructables (`new`), because of (1) and because there's no way to tell or know, other than reading the source.
4) The only scenario I can think of, where the current behaviour could apply, is when the implementation *actually* uses `this` without binding (assuming leaking implementation details are of no concern).

The linter should error on the callee. The function/method implementor should be aware of (1) and the linter should suggest implementation patterns which express clear intent and keep the call site safe from errors and workarounds to allow for (3). Even if this proves to be impossible to achieve in terms of language syntax, other than comments, it would be better to have these linter comments and errors on the callee's site (but `this: void` always works, does it?).

Edit:

A nice example for when this bites the user (ironically) is the TypeScript API. It kicks in for  `typescript.sys.fileExists` in combination with `typescript.findConfigFile` (or probably everywhere the API expects a `fileExists: (fileName: string) => boolean` callback argument. I can't imagine the author's intent is to have the user to wrap a callback into a callback, or to bind `this`. Hypothetically and perhaps far fetched one could even imagine blindly following this rule can introduce runtime errors, because the callee might rely on bad practices around the usage of this.

_Originally posted by @valler in https://github.com/typescript-eslint/typescript-eslint/issues/10056#issuecomment-2467053331_

**Comment by bradzacher** [comment]
The rule errors at the call site because a method is not unbound unless it's used in an unbound way.

```ts
const foo = {
  baz() {
    return this.val,
  },
  val: 0,
};
```

this code is 100% correct and safe. By itself there's nothing wrong with it.
If you call the code "safely" -- i.e. `foo.baz()` then the code is still correct and safe.

It's not until you do something "unsafe" eg `const x = foo.baz;` that the method becomes unbound and the code is now broken.

This is why the rule reports at the unbinding site and not the definition - because it's the unbinding of the method that is the problem.

Now you might say "well `const x = foo.baz` isn't unsafe in of itself" and you're not wrong. Technically a reassignment could be followed by `x.call(foo)` which would be a safe usage. However we don't track this because it's a hard problem to solve.
The types do not track the origin of the function (i.e. if we inspect the type of `x` we just would see the type `() => void` and nothing more). So if we wanted to track the eventual call of `x` to determine that if it was called correctly then we'd need to implement full scope analysis to track the flow of the value through branches and such to determine this. That's expensive and hard to do for even the simplest cases and it's ultimately not worth it as you won't really improve the experience of the rule -- it's very rare that people do things like "assign a method to later bind it correctly" so it's not worth the complexity and cost.

**Comment by bradzacher** [comment]
Note that the rule "understands" the difference between arrow functions and function declarations/expressions in that arrow functions permanently bind their `this` context at declaration time and thus can be safely used in way that would otherwise unbind a function declaration/expression.

So this is why usages of `bee` in your playground do not error like usages of `bar` and `baz` do - because the type of `foo` declares `bee: () => void` -- so the rule determines that that is implemented as an arrow function and thus cannot be unbound.

So for the rule:
```ts
type T = {
  method(): void; // can be unbound
  arrowProp: () => void; // cannot be unbound
};

const obj = {
  method() { }, // can be unbound
  funcProp: function () { }, // can be unbound
  arrowProp: () => { }, // cannot be unbound
};

class Clazz {
  method() { }, // can be unbound
  funcProp = function () { }, // can be unbound
  arrowProp = () => { }, // cannot be unbound
}
```

**Comment by bradzacher** [comment]
It also looks like you're looking at this from the perspective of "the rule should be smart enough to understand whether or not a function uses `this` so it can understand whether or not there is a `this` context to care about.

In an ideal world that would be great.
But that again requires deep complex analysis of the code that we simply cannot do.
In order to do this we'd need to analyse every single relevant function and method in a the codebase to try to understand this -- which is a huge amount of work to be performing.

Additionally TypeScript does not encode this information in types -- so we can't rely on TS for this. 

You as a user could annotate this information in your code if you so desired by way of a `this: void` parameter which the rule will respect. But most people don't like to do this because it's verbose.

**Comment by valler** [comment]
Thanks @bradzacher 

I am aware of the differences to arrow functions and when unbinding happens. I added bee to illustrate (3) of my additional info. And yes, I was exactly trying to make the point, that the linter should be able to infer `this` usage **or** hint at `this:void` usage. Unbound methods are no less safe than bound methods, unless they use `this`. That's why, I would expect no error for `bar`, even when unbinding from `foo`.

Second, unbinding functions/methods should be preferred over wrapping/binding to pass callbacks. The user's desire for clean and simple code should outweigh the implementors desire to avoid `this: void` or arrow functions, because the implementation happens *once*, while usage is *unlimited*. Without knowledge of the implementation a user simply has no way to tell wether unbinding is safe, thus requiring the user to bind/wrap every single callback passed into an API, unless he would be willing to perform the code analysis the linter currently refuses to do (maybe for reasons).

**Comment by bradzacher** [comment]
> Second, unbinding functions/methods should be preferred over wrapping/binding to pass callbacks

That's your stylistic opinion -- and one that most people disagree with.
Wrapping functions with an arrow function to preserve the `this` is considered idiomatic in modern JS.

---

> unless he would be willing to perform the code analysis the linter currently refuses to do

I mentioned this above - we're unable to analyse methods to determine if they use `this`.
Unfortunately this is just a limitation of the rule and linting in general.

It's technically possible -- but there is huge complexity involved in solving this problem in the general case. Additionally the performance hit required to do this across an entire codebase would make the rule unusable for most people.

As such it's not something we'll build into the rule.

The rule stops at the signature and that's as much information as it can gather and infer about a method.

The options for avoiding reports from the rule:

1) use `this: void`
2) use arrow functions at the declaration site
3) wrap the usage in a function expression to avoid unbinding
4) `.bind` / `.call` to preserve the `this`.

**Comment by valler** [comment]
> That's your stylistic opinion -- and one that most people disagree with.
> Wrapping functions with an arrow function to preserve the `this` is considered idiomatic in modern JS.

Indeed just an opinion. I would not consider wrapping idiomatic, because of the demonstrated code bloat. If that's the direction the taken, all I can do is hope devs make informed choices about when to prefer arrow functions over syntactic sugar for methods. As pointed out, not even the Typescript API currently does so.
